### ISSUE TITLE
B2 Windows: corrupt extfs desktop icon

### ISSUE THREAD

**Issue by rakslice** [rakslice/macemu#10]
In Basilisk II, under Windows, the top level extfs volume that appears on the desktop shows up as just named "T". 

Under Mac OS 8 it has the old-style black & white hard drive icon appearance initially, but it goes corrupted when it is selected.

**Comment by rakslice** [comment]
This hard drive icon was the state saved in the Virtual Desktop folder on the host from a different version; however the name and the corrupt icon after selecting persist after removing the state

**Comment by rakslice** [comment]
The changes to `user_strings_windows.cpp` in a6a46a269756023e5d816b2f6597cc22510246da did not take into account that switching to the unicode Windows registry API would cause `RegQueryValueEx` to provide string values as wchar_t strings.

**Comment by rakslice** [comment]
However when I made code changes to fix this, I started having illegal memory accesses when extfs first loads.

Is this a direct result of my change somehow? Or did was the different return value from `get_volume_name()` mask some other problem?

**Comment by rakslice** [comment]
The thing that's in progress during the crash is `fs_get_file_info` on the `Icon\r` file created for the virtual desktop.

I see that the main "data fork" `Icon\r` file has tremendously old modified and created time values (Aug 1 1930, Feb 19 1965, respectively) according to both MSYS shell tools and Windows itself. Actually the `stat` result coming back is -1 for both mtime and ctime; perhaps that's what it is?
The corresponding `.rsrc` and `.finif` files for the `Icon\r` file have the time of the `Virtual Desktop` creation, as expected.

**Comment by rakslice** [comment]
So `TimeToMacTime` calls `localtime(-1)` which apparently is bad news on Windows.

**Comment by rakslice** [comment]
The weird thing is that statting `Icon\r` right after it is created with the underlying `_tcreat` has normal current times. It is only after `posix_emu` has done further stuff to it that it has null times.

```
[47092] B2: 02/12/20 05:19:48:066 creat(Icon ,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D,0) = 3
[47092] B2: 02/12/20 05:19:48:066 close(3) = 0
[47092] B2: 02/12/20 05:19:48:066 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon )...
[47092] B2: 02/12/20 05:19:48:066 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D
[47092] B2: 02/12/20 05:19:48:066 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon )...
[47092] B2: 02/12/20 05:19:48:066 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D
[47092] B2: 02/12/20 05:19:48:066 stat(Icon ,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D) = 0
[47092] B2: 02/12/20 05:19:48:066 stat: size = 0, mode = 33206, a = 1581513588, m = 1581513588, c = 1581513588
[47092] B2: 02/12/20 05:19:48:066 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.rsrc\Icon )...
[47092] B2: 02/12/20 05:19:48:066 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.rsrc\Icon%0D
[47092] B2: 02/12/20 05:19:48:067 open-creat(.rsrc\Icon ,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.rsrc\Icon%0D,258) = 3
[47092] B2: 02/12/20 05:19:48:067 write(3,00742d40,2670) = 2670
[47092] B2: 02/12/20 05:19:48:068 close(3) = 0
[47092] B2: 02/12/20 05:19:48:068 get_finfo on Icon 
[47092] B2: 02/12/20 05:19:48:068 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\Icon )...
[47092] B2: 02/12/20 05:19:48:068 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\Icon%0D
[47092] B2: 02/12/20 05:19:48:068 open(.finf\Icon ,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\Icon%0D,0) = -1
[47092] B2: 02/12/20 05:19:48:069 set_finfo(Icon , -53700284, 0, 0)
[47092] B2: 02/12/20 05:19:48:069 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon )...
[47092] B2: 02/12/20 05:19:48:069 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D
[47092] B2: 02/12/20 05:19:48:069 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\Icon )...
[47092] B2: 02/12/20 05:19:48:069 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\Icon%0D
[47092] B2: 02/12/20 05:19:48:070 open-creat(.finf\Icon ,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\Icon%0D,258) = 3
[47092] B2: 02/12/20 05:19:48:070 write(3,00809944,16) = 16
[47092] B2: 02/12/20 05:19:48:071 close(3) = 0
[47092] B2: 02/12/20 05:19:48:072 get_finfo on This PC
[47092] B2: 02/12/20 05:19:48:072 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\This PC)...
[47092] B2: 02/12/20 05:19:48:072 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\This PC
[47092] B2: 02/12/20 05:19:48:072 open(.finf\This PC,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\This PC,0) = -1
[47092] B2: 02/12/20 05:19:48:072 set_finfo(This PC, -53700284, 0, 1)
[47092] B2: 02/12/20 05:19:48:072 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\This PC)...
[47092] B2: 02/12/20 05:19:48:072 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\This PC
[47092] B2: 02/12/20 05:19:48:072 utime failed on This PC, error 2
[47092] B2: 02/12/20 05:19:48:072 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\This PC)...
[47092] B2: 02/12/20 05:19:48:072 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\This PC
[47092] B2: 02/12/20 05:19:48:073 open-creat(.finf\This PC,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\.finf\This PC,258) = 3
[47092] B2: 02/12/20 05:19:48:073 write(3,00809944,16) = 16
[47092] B2: 02/12/20 05:19:48:074 close(3) = 0
[47092] B2: 02/12/20 05:19:48:074 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon )...
[47092] B2: 02/12/20 05:19:48:074 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D
[47092] B2: 02/12/20 05:19:48:074 charset_mac2host(C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon )...
[47092] B2: 02/12/20 05:19:48:074 charset_mac2host = C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D
[47092] B2: 02/12/20 05:19:48:074 stat(Icon ,C:\msys32\home\Andrew Tonner\src\fresh_test\Virtual Desktop\Icon%0D) = 0
[47092] B2: 02/12/20 05:19:48:074 stat: size = 0, mode = 33206, a = 1581513588, m = -1, c = -1
```

**Comment by rakslice** [comment]
So to init the finfo the Mac expects for an `Icon\r`, instead of just creating the `.finf` file itself, the representation it itself expects, `posix_emu`'s little routine for setting up the icon calls `get_finfo`, twiddles the struct a bit, then does `set_finfo`.
`get_finfo`, on the absence of an actual `.finf` to read, is mostly serving up a zeroed `SIZEOF_FInfo` block of bytes.

But `set_finfo` has a trick up it's sleeve: 
In addition to writing the `.finf` block for the finfo data it is given, it also reads backwards (into the Hinfo?) to get the file times, and writes them to the main file.

```
void set_finfo(const char *path, uint32 finfo, uint32 fxinfo, bool is_dir)
{
        D(bug("set_finfo(%s, %d, %d, %d)\n", path, finfo, fxinfo, is_dir));
        struct my_utimbuf times;
        times.actime = MacTimeToTime(ReadMacInt32(finfo - ioFlFndrInfo + ioFlCrDat));
        times.modtime = MacTimeToTime(ReadMacInt32(finfo - ioFlFndrInfo + ioFlMdDat));

        if (utime(path, &times) < 0) {
                D(bug("utime failed on %s, error %d\n", path, GetLastError()));
        }
```

If this codebase has much stuff like this around, no wonder it doesn't work with proper memory segmentation handling.

**Comment by rakslice** [comment]
07e9db881c109ad060611f974e2d92402ca593af
837c82d8b893804dc5242ecc378c6a228f883d03
8b44b00da3449164e9a5409ae45223ec989253b7

**Comment by rakslice** [comment]
I think the corrupt icon background when selected is specific to Mac OS 8.x, and a separate issue for now
